pkg/controllers/workspace_controller.go

```diff
@@ -168,13 +173,20 @@ func selectWorkspaceNodes(qualified []*corev1.Node, preferred []string, previous
 			} else if !iPrevious && jPrevious {
 				return false
 			} else { // either all are previous, or none is previous
-				_, iCreatedByKaito := qualified[i].Labels["kaito.sh/machine-type"]
-				_, jCreatedByKaito := qualified[j].Labels["kaito.sh/machine-type"]
-
-				// Choose node created by gpu-provisioner since it is more likely to be empty to use.
-				if iCreatedByKaito && !jCreatedByKaito {
+				var iCreatedByGPUProvisioner, jCreatedByGPUProvisioner bool
+				_, iCreatedByGPUProvisioner = qualified[i].Labels[machine.LabelGPUProvisionerCustom]
+				_, jCreatedByGPUProvisioner = qualified[j].Labels[machine.LabelGPUProvisionerCustom]
+				// Choose node created by gpu-provisioner and karpenter since it is more likely to be empty to use.
+				var iCreatedByKarpenter, jCreatedByKarpenter bool
+				if featuregates.FeatureGates[consts.FeatureFlagKarpenter] {
+					_, iCreatedByKarpenter = qualified[i].Labels[nodeclaim.LabelNodePool]
+					_, jCreatedByKarpenter = qualified[j].Labels[nodeclaim.LabelNodePool]
+				}
+				if (iCreatedByGPUProvisioner && !jCreatedByGPUProvisioner) ||
+					(iCreatedByKarpenter && !jCreatedByKarpenter) {
```

Comment:
So this is handling the selection criteria of eligible nodes if there are both gpu-provisioner and karpenter nodes?

Reply:
Oh I see, all good